### PR TITLE
Initial implementation of salesforcewaveanalytics

### PR DESCRIPTION
**Pull request by Nirubikaa**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,38 @@
...
+# Salesforce Wave Analytics EI Connector
+
+The Salesforce Wave Analytics [Connector](https://docs.wso2.com/display/EI640/Working+with+Connectors) allows you to work with records in Salesforce. You can use the Salesforce connector to create, query, retrieve and update records in your organization's Salesforce data. The connector uses the [Analytics REST API](https://developer.salesforce.com/docs/atlas.en-us.bi_dev_guide_rest.meta/bi_dev_guide_rest/bi_rest_overview.htm) to interact with Salesforce.
+
+
+## Compatibility
+
+| Connector version | Supported Salesforce REST API version | Supported WSO2 ESB/EI version |
+| ----------------- | ------------- | ------------- |
+| [1.0.0](https://github.com/wso2-extensions/esb-connector-salesforcewaveanalytics/tree/org.wso2.carbon.connector.salesfroceWaveAnalytics-1.0.0) | v32.0 to v45.0 | EI 6.4.0 |
```
Comment (Biruntha) [inline README.md]:
```suggestion
| [1.0.0](https://github.com/wso2-extensions/esb-connector-salesforcewaveanalytics/tree/org.wso2.carbon.esb.connector.salesfroceWaveAnalytics-1.0.0) | v32.0 to v45.0 | EI 6.4.0 |
```

File: docs/config.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+   
+        i.  Export the Salseforce certificate "DigiCertSHA2SecureServerCA.crt" form the salesforce home page.
+           
+        ii. Open a command line terminal and execute the following command to import the certificate to  client-truststore.jks and wso2carbon.jks:
+               "keytool -importcert -file DigiCertSHA2SecureServerCA.crt -keystore {SALESFORCEWAVEANALYTICS_CONNECTOR_HOME}/src/test/resources/keystores/products/client-truststore.jks -alias salesforce" and "keytool -importcert -file DigiCertSHA2SecureServerCA.crt -keystore {SALESFORCEWAVEANALYTICS_CONNECTOR_HOME}/src/test/resources/keystores/products/wso2carbon.jks -alias salesforce"
+           
+
+Salesforce Wave Analytics API uses the OAuth protocol to allow application users to securely access data without having to reveal their user credentials.  For more information on authentication is done in Salesforce, see [Understanding Authentication](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_understanding_authentication.htm).
+
+###Obtaining user credentials
```
Comment (Biruntha) [inline docs/config.md]:
```suggestion
### Obtaining user credentials
```

File: repository/esb-connector-salesforcewaveanalytics.properties [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+clientId=3MVG9pe2TCoA1Pf6K584VNDD80a6UqXVrD8cmFybYmBGw_0lzlTDE.O6.jp8U4Dnlw6WKH62Rwp7DAHjnd7sl
+accessToken=00D0o0000016avV!AQYAQEWJpklFd8cgy0AYtQWa0b059dgcm.7Mr2QJweqfJcOchQ5z864PRp_DgnLDo9vwqNA7UrDt4b7qQjSSof42MkhEL.oz
+hostName=https://login.salesforce.com
+apiUrl=https://ap8.salesforce.com
+intervalTime=2400000
+registryPath=connectors/Salesforcewaveanalytics
+
+
+attachedFilesId=0690o000005u2XH
+
```
Comment (Biruntha) [inline repository/esb-connector-salesforcewaveanalytics.properties]: Remove empty lines

File: src/main/java/org/wso2/carbon/esb/connector/salesforceWaveAnalyticsConnector.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        Object templateParam = getParameter(messageContext, "generated_param");
+        try {
+            log.info("salesforcewaveanalytics sample connector received message :" + templateParam);
+            /**Add your connector code here 
+            **/
+        } catch (Exception e) {
+	    throw new ConnectException(e);	
+        }
+    }
+}
```
Comment (Biruntha) [inline src/main/java/org/wso2/carbon/esb/connector/salesforceWaveAnalyticsConnector.java]: What is the purpose of this class? If not used , pls remove

File: src/main/resources/config/init.xml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        <property name="uri.var.clientSecret" expression="$func:clientSecret"/>
+        <property name="uri.var.clientId" expression="$func:clientId"/>
+        <property name="uri.var.apiUrl" expression="$func:apiUrl"/>
+        <property name="uri.var.registryPath" expression="$func:registryPath"/>
+        <property name="uri.var.intervalTime" expression="$func:intervalTime"/>
+        <property name="timeout" expression="$func:timeout"/>
+        <script language="js">
+            <![CDATA[
+                var timeout = mc.getProperty("timeout");
+                if (timeout !=null && timeout !=""){
```
Comment (Biruntha) [inline src/main/resources/config/init.xml]:
```suggestion
                if (timeout !=null && timeout !="") {
```

File: src/main/resources/config/init.xml [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                               mc.getConfiguration().getRegistry().updateResource(newAccessRegistryPath, newAccess);
+                               mc.getConfiguration().getRegistry().newResource(newTimeRegistryPath, false);
+                               mc.getConfiguration().getRegistry().updateResource(newTimeRegistryPath, newSysTime);
+                           }
+                           ]]>
+                        </script>
+                    </then>
+                    <else>
+                        <property name="uri.var.accessToken"
+                                  expression="$ctx:uri.var.accessToken.reg"/>
```
Comment (Biruntha) [inline src/main/resources/config/init.xml]: Move this line to above

File: src/main/resources/dashboard/listDashboard.xml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        <property name="uri.var.pageSize" expression="$func:pageSize"/>
+        <property name="uri.var.q" expression="$func:q"/>
+        <property name="uri.var.sort" expression="$func:sort"/>
+
+        <property name="Accept-Encoding" scope="transport" action="remove"/>
+
+        <filter xpath="(not(get-property('uri.var.folderId') = '' or (not(string(get-property('uri.var.folderId'))))))">
+            <then>
+                <property name="uri.var.apiParameters"
+                          expression="fn:concat(get-property('uri.var.apiParameters'),'&amp;folderId=',get-property('uri.var.folderId'))"/>
```
Comment (Biruntha) [inline src/main/resources/dashboard/listDashboard.xml]: Exceeded 120 line limit. Check through the file

File: src/main/resources/query/query.xml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+   under the License.
+-->
+<template name="query" xmlns="http://ws.apache.org/ns/synapse">
+    <parameter name="query" description="The query, in JSON format"/>
+    <sequence>
+        <property name="query" expression="$func:query"/>
+        <payloadFactory media-type="json">
+            <format>
+                {
+                "query": $1,
```
Comment (Biruntha) [inline src/main/resources/query/query.xml]:
```suggestion
                "query": $1
```

File: src/test/java/org/wso2/carbon/connector/integration/test/salesforcewaveanalytics/SalesforceWaveAnalyticsConnectorIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+        String methodName = "listTopLevelResources";
+        RestResponse<JSONObject> esbRestResponse = sendJsonRestRequest(getProxyServiceURLHttp(methodName),
+                "POST", esbRequestHeadersMap, "listTopLevelResourcesNegative.json");
+        Assert.assertEquals(esbRestResponse.getHttpStatusCode(), 404);
+    }
+
+    /**
+     * Positive test case for listAnnotations method with mandatory parameters.
+     */
+
```
Comment (Biruntha) [inline src/test/java/org/wso2/carbon/connector/integration/test/salesforcewaveanalytics/SalesforceWaveAnalyticsConnectorIntegrationTest.java]: Remove empty line. check in all the places

File: src/test/java/org/wso2/carbon/connector/integration/test/salesforcewaveanalytics/SalesforceWaveAnalyticsConnectorIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+    @Test(groups = {"wso2.esb"}, dependsOnMethods = {"testListDashboardMandatory"},
+            description = "salesforcewaveanalytics {listAnnotations} integration test with mandatory parameters.")
+    public void testListAnnotationsMandatory() throws Exception {
+
+        String methodName = "listAnnotations";
+        RestResponse<JSONObject> esbRestResponse = sendJsonRestRequest(getProxyServiceURLHttp(methodName),
+                "POST", esbRequestHeadersMap,
+                "listAnnotationsMandatory.json");
```
Comment (Biruntha) [inline src/test/java/org/wso2/carbon/connector/integration/test/salesforcewaveanalytics/SalesforceWaveAnalyticsConnectorIntegrationTest.java]: Move this line to above. Check in all places

**Review by keerthu — APPROVED** [review]

**Review by kesavany — APPROVED** [review]

**Pull request merged**